### PR TITLE
Improve type argument list parsing in error conditions.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/19456

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-compiler @gafter @jcouv 

Note: this likely won't meet the bar for 15.3.   However, i would still like it reviewed and ready to go into the branch we open for post 15.3 work.

**Comment by jaredpar** [comment]
> Note: this likely won't meet the bar for 15.3. However, i would still like it reviewed and ready to go into the branch we open for post 15.3 work.

If this doesn't meet the bar for 15.3 then we will likely de-prioritize the review as well.  Being asked to review a lot right now for 15.3 work.  This would need to come after.

**Comment by CyrusNajmabadi** [comment]
That's fine.  I have a bunch of other PRs waiting on reviews as well.  But i still want them out there and ready for review when people have a chance.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -5275,11 +5275,39 @@ private ScanTypeArgumentListKind ScanTypeArgumentList(NameOptions options)
...
+            // This is absolutely a type-argument-list.  Just return that result immediately
+            // 
+            //      result != ScanTypeFlags.NotType && !isDefinitelyTypeArgumentList.
+            //
+            // This could be a type-argument list, or it could be an expression.  Need to see
+            // what came after the last '>' to find out which it is.
+
+            SyntaxToken lastTokenOfList = null;
+            ScanTypeFlags possibleTypeArgumentFlags = ScanPossibleTypeArgumentList(
+                ref lastTokenOfList, out bool isDefinitelyTypeArgumentList);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: note: i'm not a huge fan of having multiple results.  But i couldn't find a great way to thread the bool through the result value.  If anyone has any suggestions on the best way to do this, i'm all ears.  Specifically, i didn't want people to have to try to figure out things like "PossibleGenericTypeOrMethod" and "MustBeGenericTypeOrMethod".  But maybe that's appropriate?

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by CyrusNajmabadi** [comment]
This has been rebased to 15.6.  @dotnet/roslyn-compiler review only when convenient for you.  Note: this helps with IDE scenarios around 'add using' with partially written generics.

**Comment by CyrusNajmabadi** [comment]
retest windows_determinism_prtest  please

**Comment by CyrusNajmabadi** [comment]
@jcouv @gafter @dotnet/roslyn-compiler please take a look.  This improves a lot of IDE scenarios when there is partially written generic code.  Thanks!

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -5298,6 +5323,49 @@ private ScanTypeFlags ScanPossibleTypeArgumentList(ref SyntaxToken lastTokenOfLi
                             return ScanTypeFlags.NotType;
 
                         case ScanTypeFlags.MustBeType:
+                            // We're currently scannling a possible type-argument list.  But we're
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >scannling [](start = 47, length = 9)

s/scannling/scanning/

File: src/Compilers/CSharp/Test/Syntax/Parsing/TypeArgumentListParsingTests.cs [hunk 3]
```diff
@@ -0,0 +1,1157 @@
...
+    public class TypeArgumentListParsingTests : ParsingTests
+    {
+        public TypeArgumentListParsingTests(ITestOutputHelper output) : base(output) { }
+
+        protected override SyntaxTree ParseTree(string text, CSharpParseOptions options)
+        {
+            return SyntaxFactory.ParseSyntaxTree(text, options: options);
+        }
+
+        private ExpressionSyntax ParseExpression(string text, ParseOptions options = null)
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/TypeArgumentListParsingTests.cs]: This method seems unused.

File: src/Compilers/CSharp/Test/Syntax/Parsing/TypeArgumentListParsingTests.cs [hunk 4]
```diff
@@ -0,0 +1,1157 @@
...
+
+        [Fact, WorkItem(19456, "https://github.com/dotnet/roslyn/issues/19456")]
+        public void TestTupleType()
+        {
+            UsingTree(@"
+class C
+{
+    void M()
+    {
+        // Currently not supported, but might be nice to have in the future.
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/TypeArgumentListParsingTests.cs]: Maybe file an issue?

File: src/Compilers/CSharp/Test/Syntax/Parsing/TypeArgumentListParsingTests.cs [hunk 5]
```diff
@@ -0,0 +1,1157 @@
...
+
+        [Fact, WorkItem(19456, "https://github.com/dotnet/roslyn/issues/19456")]
+        public void TestGenericArg()
+        {
+            UsingTree(@"
+class C
+{
+    void M()
+    {
+        var added = ImmutableDictionary<T<S>, IImmutableDictionary<X, Y>>
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/TypeArgumentListParsingTests.cs]: Shouldn't the comma (following `T<S>` give us a strong hint of type argument list here?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Fixed.

File: src/Compilers/CSharp/Test/Syntax/Parsing/TypeArgumentListParsingTests.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/TypeArgumentListParsingTests.cs]: Fixed

File: src/Compilers/CSharp/Test/Syntax/Parsing/TypeArgumentListParsingTests.cs (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/TypeArgumentListParsingTests.cs]: I'm not a fan of speculative issues like "perhaps we could do better here".  I'd rather have this driven by actual use cases.  The purpose of the comment was more to say that we're not trying to write a test stating that this must be the behavior.  rather the test just documents the existing behavior, and if we get better in the future that's totally fine.

File: src/Compilers/CSharp/Test/Syntax/Parsing/TypeArgumentListParsingTests.cs (hunk 5 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/TypeArgumentListParsingTests.cs]: Yes.  I've updated our checks to handle this.

File: src/Compilers/CSharp/Test/Syntax/Parsing/TypeArgumentListParsingTests.cs (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/TypeArgumentListParsingTests.cs]: I've updated the PR to handle tuples as well.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParsingTests.cs [hunk 6]
```diff
@@ -1,5 +1,7 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+#define PARSING_TESTS_DUMP
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParsingTests.cs]: Undo this line

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParsingTests.cs]: Done.

**Comment by CyrusNajmabadi** [comment]
Test failure is https://github.com/dotnet/roslyn/issues/15497

**Comment by CyrusNajmabadi** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler Can i get a second pair of eyes?  Note: i recommend taking a look at the test cases to see what's being checked for here.

**Comment by CyrusNajmabadi** [comment]
Ping @dotnet/roslyn-compiler

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 7]
```diff
@@ -5279,6 +5304,70 @@ private ScanTypeFlags ScanPossibleTypeArgumentList(ref SyntaxToken lastTokenOfLi
...
+                            //
+                            // Note: we check if we got 'MustBeType' which triggers for predefined types,
+                            // (int, string, etc.), or array types (Foo[], A<T>[][] etc.), or pointer types
+                            // of things that must be types (int*, void**, etc.).
+                            isDefinitelyTypeArgumentList = DetermineIfDefinitelyTypeArgumentList(isDefinitelyTypeArgumentList);
+                            result = ScanTypeFlags.GenericTypeOrMethod;
+                            break;
+
+                        case ScanTypeFlags.TupleType:
+                            // See above.  If we have  X<(a, b),   or  X<(a, b)> then this is definitely a type argument list.
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: See `X<(a, b),` below. Is this a situation that this code would disambiguate incorrectly?

``` c#
public class C
{
    public static void Main()
    {
        XX X;
        int a = 1, b = 2;
        bool z = X < (a, b), w = false;
    }
}

struct XX
{
    public static bool operator <(XX x, (int a, int b) arg) => true;
    public static bool operator >(XX x, (int a, int b) arg) => false;
}
```

Comment (gafter) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: We should have a test covering this situation, please.

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Nope.  That's a good catch.  I will fix up the check, and also add a test to make sure we don't mess this up.

**Comment by CyrusNajmabadi** [comment]
PR updated with test.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParsingTests.cs [hunk 8]
```diff
@@ -1,5 +1,7 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+// #define PARSING_TESTS_DUMP
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParsingTests.cs]: This may be the most useful comment ever (seriously).

Comment (gafter) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParsingTests.cs]: Though an explanation accompanying it would be helpful.

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParsingTests.cs]: agreed :)

**Pull request merged**